sympy/solvers/ode/single.py

```diff
@@ -791,66 +791,101 @@ def _get_general_solution(self, *, simplify: bool = True):
         return [gensol]
 
 
-class SecondNonlinearAutonomousConserved(SinglePatternODESolver):
+class SecondAutonomousNth(SinglePatternODESolver):
     r"""
-    Gives solution for the autonomous second order nonlinear
-    differential equation of the form
+    The general second order autonomous differential equation of the form:
 
-    .. math :: f''(x) = g(f(x))
+    .. math :: x''(t) = g(x'(t), x(t))
 
-    The solution for this differential equation can be computed
-    by multiplying by `f'(x)` and integrating on both sides,
-    converting it into a first order differential equation.
+    This solver is added for the special case in the above genre:
+
+    .. math :: f''(x) = g(f(x)) f'(x)^n
+
+    Explanation
+    ===========
+
+    This special case can be solved using by writing the second derivative in
+    form involving a power of f'(x) and on rewriting the second derivative, then rearranging.
+    This solver works for any arbitrary value of n defined above.
+
+    Extras: For faster results, kindly set simplify = False in dsolve().
 
     Examples
     ========
 
     >>> from sympy import Function, symbols, dsolve
     >>> f, g = symbols('f g', cls=Function)
-    >>> x = symbols('x')
+    >>> x, eq, n = symbols('x eq n')
 
-    >>> eq = f(x).diff(x, 2) - g(f(x))
-    >>> dsolve(eq, simplify=False)
-    [Eq(Integral(1/sqrt(C1 + 2*Integral(g(_u), _u)), (_u, f(x))), C2 + x),
-    Eq(Integral(1/sqrt(C1 + 2*Integral(g(_u), _u)), (_u, f(x))), C2 - x)]
+    >>> eq = f(x).diff(x, x) + f(x).diff(x)**2
+    >>> dsolve(eq, hint='2nd_autonomous_nth')
+    Eq(f(x), log(C1*(C2 + x)))
 
-    >>> from sympy import exp, log
+    >>> eq = f(x).diff(x, x) - f(x).diff(x)**2
+    >>> dsolve(eq, hint='2nd_autonomous_nth')
+    Eq(f(x), log(C1/(C2 + x)))
+
+    >>> from sympy import exp, log, cbrt
     >>> eq = f(x).diff(x, 2) - exp(f(x)) + log(f(x))
-    >>> dsolve(eq, simplify=False)
-    [Eq(Integral(1/sqrt(-2*_u*log(_u) + 2*_u + C1 + 2*exp(_u)), (_u, f(x))), C2 + x),
-    Eq(Integral(1/sqrt(-2*_u*log(_u) + 2*_u + C1 + 2*exp(_u)), (_u, f(x))), C2 - x)]
+    >>> dsolve(eq, hint='2nd_autonomous_nth')
+    [Eq(Integral(1/sqrt(-2*_v*log(_v) + 2*_v + C1 + 2*exp(_v)), (_v, f(x))), C2 + x),
+    Eq(-Integral(1/sqrt(-2*_v*log(_v) + 2*_v + C1 + 2*exp(_v)), (_v, f(x))), C2 + x)]
+
+    >>> eq = f(x).diff(x, 2) + cbrt(f(x)) + 1/f(x)
+    >>> dsolve(eq, hint='2nd_autonomous_nth')
+    [Eq(sqrt(2)*Integral(1/sqrt(-3*_v**(4/3) + C1 - 4*log(_v)), (_v, f(x))), C2 + x),
+    Eq(-sqrt(2)*Integral(1/sqrt(-3*_v**(4/3) + C1 - 4*log(_v)), (_v, f(x))), C2 + x)]
+
+    >>> eq = f(x).diff(x, x) + f(x).diff(x)**4
+    >>> dsolve(eq, hint='2nd_autonomous_nth')
+    Eq((C1 + 2*f(x))**(3/2)/3, C2 + x)
```

Comment:
There do not need to be so many examples in the docstring. Only one or possibly two examples are needed and they should be chosen carefully. Often the best example is just the general form of ODE that is solved by the method.

Examples that don't include `f(x)` are not interesting as they are more efficiently solved by reduction of order.

Reply:
Sure, I removed all the examples and now it has just one example `f(x).diff(x, 2) + a*f(x).diff(x)**n`.
Will this work?
Thanks.

Reply:
That example is better solved by reduction of order i.e. the substitution `g = f'`.

Reply:
Will `f(x).diff(x, 2) + g(f(x))*f(x).diff(x)**n`?
It gives `Eq(Integral((C1 + (n - 2)*Integral(g(_v), _v))**(1/(n - 2)), (_v, f(x))), C2 + x)`.
We can't use `subs` since for different n values, there are different return types for equations.

Reply:
Can we use a Piecewise to handle the different cases?

Reply:
Something likes this works with the `subs` method:

```
def _get_general_solution(self, *, simplify: bool = True):
        a, n = self.wilds_match()
        fx = self.ode_problem.func
        x = self.ode_problem.sym
        (C1, C2) = self.ode_problem.get_numbered_constants(num=2)
        v = Dummy('v')
        a = a.subs(fx, v)
        c = C1*Integral(exp(Integral(a, v)), (v, fx))
        d = Integral((C1 + (n - 2)*Integral(a, v))**(1/(n - 2)), (v, fx))
        from sympy import Piecewise
        return Piecewise(((Eq(d, C2 + x), Eq(-d, C2 + x)), Eq(n,0)), (Eq(c, C2 + x), Eq(n, 2)), (Eq(d, C2 + x), True))
```

But this again arises the previous issue that if `n` is an undefined symbol, then set `simplify` as `False`, otherwise, it gives `ValueError` as `sympy` attempts to solve the integral, it imposes more Piecewise conditions on n in the Integral, hence it structs somehow in `odesimp` as: 

For `eq = f(x).diff(x, 2) + f(x).diff(x)**n` and `sol = dsolve(eq, hint='2nd_autonomous_nth')`:
```
<ipython-input-2-a6569a07ae34> in <module>
----> 1 sol = dsolve(eq, hint='2nd_autonomous_nth')

~\sympy\sympy\solvers\ode\ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    661             # The key 'hint' stores the hint needed to be solved for.
    662             hint = hints['hint']
--> 663             return _helper_simplify(eq, hint, hints, simplify, ics=ics)
    664
    665 def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):

~\sympy\sympy\solvers\ode\ode.py in _helper_simplify(eq, hint, match, simplify, ics, **kwargs)
    696             rv = [odesimp(eq, s, func, hint) for s in sols]
    697         else:
--> 698             rv =  odesimp(eq, sols, func, hint)
    699     else:
    700         # We still want to integrate (you can disable it separately with the hint)

~\sympy\sympy\core\multidimensional.py in wrapper(*args, **kwargs)
    135                     result = apply_on_element(wrapper, args, kwargs, n)
    136                     return result
--> 137             return f(*args, **kwargs)
    138         return wrapper

~\sympy\sympy\solvers\ode\ode.py in odesimp(ode, eq, func, hint)
   2269         eq = simplify(eq)
   2270     if not isinstance(eq, Equality):
-> 2271         raise TypeError("eq should be an instance of Equality")
   2272
   2273     # Second, clean up the arbitrary constants.

TypeError: eq should be an instance of Equality
```

For `eq = f(x).diff(x, 2) + f(x).diff(x)**n` and `sol = dsolve(eq, hint='2nd_autonomous_nth', simplify = False)`:
```
In [11]: print(sol)
Piecewise(((Eq(Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), C2 + x), Eq(-Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), C2 + x)), Eq(n, 0)), (Eq(C1*exp(f(x)), C2 + x), Eq(n, 2)), (Eq(Piecewise(((C1 + (n - 2)*f(x))**(1 + 1/(n - 2))/((1 + 1/(n - 2))*(n - 2)), Ne(1/(n - 2), -1)), (log(C1 + (n - 2)*f(x))/(n - 2), True)), C2 + x), True))
```

and

```
In [8]: sol.subs(n,2)
Out[8]:
    f(x)
C₁⋅ℯ     = C₂ + x

In [9]: sol.subs(n,4)
Out[9]:
             3/2
(C₁ + 2⋅f(x))
──────────────── = C₂ + x
       3

In [10]: sol.subs(n,1)
Out[10]: -log(C₁ - f(x)) = C₂ + x

```
So on summarising, if `n` is `undefined` then we need to set `simplify` as `False`, otherwise, we don't need to for any other case.

Reply:
> otherwise, it gives `ValueError`

That just seems like a bug somehow that could be fixed. `get_general_solution` should return a list of `Eq`.

When I suggested a Piecewise I didn't mean a Piecewise that has `Eq` as its cases. I meant a list of Eqs where one of the sides of the Eqs is a Piecewise.

Reply:
Will `[Eq(Piecewise((c, Eq(n, 2)), ((d, -d), Eq(n, 0)), (d, True)), C2+x)]` this work?
But this gives us `AttributeError: 'Tuple' object has no attribute 'as_coeff_Mul'`, how can we return two solutions for `n==0`.
Thanks.

Reply:
Yes, it should be something like that. It should be possible to structure it so that you always return two Eqs but substituting some values of `n` makes them equal.

Reply:
This should work `[Eq(Piecewise((d, Eq(n, 0)), (c, Eq(n, 2)), (d, True)), C2+x), Eq(Piecewise((-d, Eq(n, 0)), (c, Eq(n, 2)), (d, True)), C2+x)]`, but somehow it gives `AttributeError: 'list' object has no attribute 'rhs'` in `_remove_redundant_solutions`, though the idea is for `n==0` it will give pair of solutions and for other cases it will find both Equations to be same, thus, `_remove_redundant_solutions` should remove one of the solutions.

Reply:
Returning this `[Eq(Piecewise((Piecewise((e,Eq(n,0)),(f,Eq(n, 1)),(d, True)), Ne(n, 2)), (nan, True)), C2+x), Eq(Piecewise((-e, Eq(n, 0)), (c, Eq(n, 2)), (nan, True)),C2+x)]` raises `RecursionError`, but no `AttributeError` as before.
I tried a lot of combinations to fix it, but nothing worked out for me, the issue is that `sympy` solves the integral further and imposes piecewise conditions on `n`, thus this makes `solns` to be` list of lists`, thus `_remove_redundant_solutions` doesn't work properly as it is supposed to have list of equations.